internal/provider/resource_vault_plugin.go

```diff
@@ -0,0 +1,322 @@
+The ID of the HCP project where the HCP Vault cluster is located. 
+If not specified, the project specified in the HCP Provider config block will be used, if configured.
+If a project is not configured in the HCP Provider config block, the oldest project in the organization will be used.`,
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+				Computed:     true,
+			},
+		},
+	}
+}
+
+func resourceVaultPluginCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client)
+
+	clusterID := d.Get("cluster_id").(string)
+	projectID, err := GetProjectID(d.Get("project_id").(string), client.Config.ProjectID)
+	if err != nil {
+		return diag.Errorf("unable to retrieve project ID: %v", err)
+	}
+	pluginName := d.Get("plugin_name").(string)
+	pluginType := d.Get("plugin_type").(string)
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      projectID,
+	}
+
+	log.Printf("[INFO] Adding Vault Plugin (%s) on Vault Cluster (%s)", pluginName, clusterID)
+
+	req := &vaultmodels.HashicorpCloudVault20201125AddPluginRequest{PluginName: pluginName, PluginType: pluginType}
+	_, err = clients.AddPlugin(ctx, client, loc, clusterID, req)
+	if err != nil {
+		return diag.Errorf("error adding plugin (%s) to Vault cluster (%s): %v", pluginName, clusterID, err)
+	}
+
+	d.SetId(vaultPluginResourceID(projectID, clusterID, pluginType, pluginName))
+
+	if err := setVaultPluginResourceData(d, projectID, clusterID, pluginName, pluginType); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceVaultPluginRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client)
+
+	idParts := strings.SplitN(d.Id(), "/", 8)
+
+	clusterID := idParts[4]
+	projectID := idParts[2]
+	pluginName := idParts[7]
+	pluginTypeString := idParts[6]
+	pluginType := vaultmodels.HashicorpCloudVault20201125PluginType(pluginTypeString)
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      projectID,
+	}
+
+	log.Printf("[INFO] Listing plugins for Vault cluster (%s) [project_id=%s, organization_id=%s]", clusterID, loc.ProjectID, loc.OrganizationID)
+
+	pluginsResp, err := clients.ListPlugins(ctx, client, loc, clusterID)
+	if err != nil {
+		log.Printf("[ERROR] Vault cluster (%s) failed to list plugins", clusterID)
+		return diag.FromErr(err)
+	}
+
+	found := false
+	for _, plugin := range pluginsResp.Plugins {
+		if strings.EqualFold(pluginName, plugin.PluginName) && pluginType == *plugin.PluginType && plugin.IsRegistered {
+			// Cluster found, update resource data.
+			if err := setVaultPluginResourceData(d, loc.ProjectID, clusterID, pluginName, pluginTypeString); err != nil {
+				return diag.FromErr(err)
+			}
+			found = true
+			break
```

Comment:
should we return here?  I am seeing two `setVaultPluginResourceData` calls when `found=true`

Reply:
good catch, I can remove the other call. I added the other call recently after going through the best practice guide and didnt see the previous call